services/powchain/src/main/java/tech/pegasys/teku/services/powchain/execution/ExecutionEngineService.java

```diff
@@ -57,45 +63,99 @@ public void chainHeadUpdated(
-                    .map(this::onHeadBlockMerge)
-                    .orElseGet(() -> SafeFuture.completedFuture(null)))
-        .finish(err -> LOG.warn("setHead failed", err));
-  }
+    final Optional<Bytes32> curFinalizedBlock;
 
-  private SafeFuture<Void> onHeadBlockMerge(BeaconBlockBodyMerge blockBody) {
-    // Check if there is a payload
-    if (!blockBody.getExecution_payload().equals(new ExecutionPayload())) {
-      return executionEngineChannel.setHead(blockBody.getExecution_payload().getBlock_hash());
-    } else {
-      return SafeFuture.completedFuture(null);
+    synchronized (this) {
+      lastBestBlock = Optional.of(bestBlockRoot);
+      curFinalizedBlock = lastFinalizedBlock;
     }
+
+    updateForkChoice(bestBlockRoot, curFinalizedBlock);
   }
 
   @Override
   public void onNewFinalizedCheckpoint(Checkpoint checkpoint) {
-    recentChainData
-        .retrieveBlockByRoot(checkpoint.getRoot())
+    final Bytes32 curBestBlock;
+    final Optional<Bytes32> curFinalizedBlock;
+
+    synchronized (this) {
+      lastFinalizedBlock = Optional.of(checkpoint.getRoot());
+      if (lastBestBlock.isEmpty()) {
+        return;
+      }
+      curBestBlock = lastBestBlock.get();
+      curFinalizedBlock = lastFinalizedBlock;
+    }
+
+    updateForkChoice(curBestBlock, curFinalizedBlock);
+  }
+
+  private static Optional<Bytes32> getPowBlockHash(Optional<BeaconBlock> block) {
+    return block
+        .map(BeaconBlock::getBody)
+        .flatMap(BeaconBlockBody::toVersionMerge)
+        .map(BeaconBlockBodyMerge::getExecution_payload)
+        .filter(executionPayload -> !DEFAULT_EXECUTION_PAYLOAD.equals(executionPayload))
+        .map(ExecutionPayload::getBlock_hash);
+  }
+
+  private void updateForkChoice(Bytes32 bestBlockRoot, Optional<Bytes32> maybeFinalizedBlockRoot) {
+
+    SafeFuture<Optional<Bytes32>> bestBlockHashPromise =
+        recentChainData
+            .retrieveBlockByRoot(bestBlockRoot)
+            .thenApply(ExecutionEngineService::getPowBlockHash);
+
+    SafeFuture<Bytes32> finalizedBlockHashPromise =
+        maybeFinalizedBlockRoot
+            .map(recentChainData::retrieveBlockByRoot)
+            .orElseGet(() -> SafeFuture.completedFuture(Optional.empty()))
+            .thenApply(ExecutionEngineService::getPowBlockHash)
+            .thenApply(maybeHash -> maybeHash.orElse(Bytes32.ZERO));
+
+    bestBlockHashPromise
+        .thenCombine(finalizedBlockHashPromise, ForkChoiceUpdate::fromBestAndFinalized)
         .thenCompose(
-            maybeABlock ->
-                maybeABlock
-                    .flatMap(block -> block.getBody().toVersionMerge())
-                    .map(this::onFinalizedBlockMerge)
+            maybeUpd ->
+                maybeUpd
+                    .map(
+                        upd ->
+                            executionEngineChannel.forkChoiceUpdated(
+                                upd.getBestBlock(), upd.getFinalizedBlock()))
```

Comment:
Are we sure that in each case the head and the finalized blocks will be consistent with each other? What data will be send in a corner case when the most recent finalized checkpoint is from a different branch than the most recent head, just because finalized checkpoint has been updated before the head and head is pending? Is current approach save from sending inconsistent data?

Reply:
Good point. Agree that this is not an ideal solution. Let me check if we could fix this

Reply:
@mkalinin I believe this should be a better variant: c3e25e3